### PR TITLE
feat: change TokenInList to show balance

### LINKED ISSUE #2926: [Demand] Display token balance in the select token dialog
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

image

## Part 2: Problem to Solve

image

## Notes

[Uniswap Interface](https://github.com/Uniswap/uniswap-interface) is a good reference.

### PR DESCRIPTION
**Pull request by NikoManiac**
closes #MASK-51
closes #2926

### CODE REVIEW

**Comment by septs** [comment]
@NikoManiac plz resolve conflict and rename the pr title.

File: packages/dashboard/package.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
   "dependencies": {
     "@dimensiondev/holoflows-kit": "0.8.0-20210317064617-6c4792c",
     "@dimensiondev/icons": "workspace:*",
+    "@dimensiondev/kit": "^0.0.0-20210221102734-0b4a937",
```
Comment (guanbinrui) [inline packages/dashboard/package.json:16]:
```suggestion
    "@dimensiondev/kit": "=0.0.0-20210221102734-0b4a937",
```

File: packages/dashboard/src/components/TokenIcon/index.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+//#region fix icon image
+function resolveTokenIconURLs(address: string, baseURIs: string[], chainId: ChainId, logoURI?: string) {
+    const checkSummedAddress = formatEthereumAddress(address)
+
+    if (isSameAddress(constantOfChain(CONSTANTS, chainId).NATIVE_TOKEN_ADDRESS, checkSummedAddress)) {
+        return baseURIs.map((x) => `${x}/info/logo.png`)
+    }
+
+    const specialIcon = SPECIAL_ICON_LIST.find((item) => item.address === address)
```
Comment (guanbinrui) [inline packages/dashboard/src/components/TokenIcon/index.tsx:24]: `currySameAddress`

File: packages/dashboard/src/components/TokenList/TokenListItem.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                <TokenIcon address={address} name={name} />
+            </ListItemIcon>
+            <ListItemText classes={{ primary: classes.text }}>
+                <Typography className={classes.primary} color="textPrimary" component="span">
+                    <span className={classes.symbol}>{symbol}</span>
+                    <span className={classes.name}>{name}</span>
+                </Typography>
+                <Typography className={classes.secondary} color="textSecondary" component="span">
+                    {data.isAddedToken ? (
+                        <span>{formatBalance(data.balance, token.decimals)} </span>
```
Comment (guanbinrui) [inline packages/dashboard/src/components/TokenList/TokenListItem.tsx:67]:
```suggestion
                        <span>{formatBalance(data.balance, token.decimals)}</span>
```

File: packages/dashboard/src/components/TokenList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import { some } from 'lodash-es'
+import { useERC20TokensDetailed } from '../../pages/Wallets/hooks/useERC20TokensDetailed'
+import { useDashboardI18N } from '../../locales'
+
+interface IProps {
+    onSelect(asset: Asset): void
+}
+
+const checkAddedToken = (token: FungibleTokenDetailed, tokens: FungibleTokenDetailed[]) =>
+    token.type === EthereumTokenType.Native || some(tokens, (t: any) => isSameAddress(token.address, t.address))
```
Comment (guanbinrui) [inline packages/dashboard/src/components/TokenList/index.tsx:20]: `currySameAddress`

File: packages/dashboard/src/locales/zh.json [hunk 5]
```diff
@@ -41,6 +41,7 @@
     "wallets_balance_Buy": "購買",
     "wallets_balance_Swap": "交換",
     "wallets_balance_Receive": "接收",
+    "wallets_loading_token": "Loading Token",
```
Comment (guanbinrui) [inline packages/dashboard/src/locales/zh.json:44]:
```suggestion
    "wallets_loading_token": "",
```

File: packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailedFromTokenLists.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        [allTokens],
+    )
+    //#endregion
+
+    //#region create searched tokens
+    const searchedTokens = useMemo(() => {
+        if (!keyword) return allTokens
+        return [
+            ...(EthereumAddress.isValid(keyword)
+                ? allTokens.filter((token) => isSameAddress(token.address, keyword))
```
Comment (guanbinrui) [inline packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailedFromTokenLists.ts:44]: `currySameAddress`

File: packages/maskbook/src/extension/options-page/DashboardComponents/TokenIcon.tsx [hunk 7]
```diff
@@ -5,27 +5,28 @@ import {
...
+    const checkSummedAddress = formatEthereumAddress(address)
+
+    if (isSameAddress(constantOfChain(CONSTANTS, chainId).NATIVE_TOKEN_ADDRESS, checkSummedAddress)) {
+        return baseURIs.map((x) => `${x}/info/logo.png`)
     }
-    const checksummedAddress = formatEthereumAddress(address)
-    if (iconMap[checksummedAddress]) return iconMap[checksummedAddress]
-    return `${baseURI}/assets/${checksummedAddress}/logo.png`
+
+    const specialIcon = SPECIAL_ICON_LIST.find((item) => item.address === address)
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/TokenIcon.tsx:24]: `currySameAddress`

File: packages/dashboard/src/components/TokenIcon/index.tsx (hunk 2 above)
Comment (septs) [reply packages/dashboard/src/components/TokenIcon/index.tsx:24]:
```suggestion
    const specialIcon = SPECIAL_ICON_LIST.find(currySameAddress(address))
```

File: packages/dashboard/src/components/TokenList/index.tsx (hunk 4 above)
Comment (septs) [reply packages/dashboard/src/components/TokenList/index.tsx:20]: why `t` is any?

File: packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailedFromTokenLists.ts (hunk 6 above)
Comment (septs) [reply packages/dashboard/src/pages/Wallets/hooks/useERC20TokensDetailedFromTokenLists.ts:44]:
```suggestion
                ? allTokens.filter(currySameAddress(keyword))
```

**Comment by NikoManiac** [comment]
I will close this pr because the branch hard to merge.  I will fix all the comments and create new pr for jira-51.

**Pull request closed without merging**